File: python/githubauthentication/github_access_token_provider.py
import asyncio
import httpx
from typing import Any, Dict, List
from urllib.parse import urlparse
from kiota_abstractions.authentication import AccessTokenProvider, AllowedHostsValidator

from githubauthentication.github_device_code_response import GitHubDeviceCodeResponse

class GitHubAccessTokenProvider(AccessTokenProvider):
	def __init__(
			self,
			client_id: str,
			scope: str,
			allowed_hosts: List[str]= ["api.github.com"]) -> None:
		super().__init__()
		self._client_id = client_id
		self._scope = scope
		self._allowed_hosts_validator = AllowedHostsValidator(allowed_hosts)

	def get_allowed_hosts_validator(self) -> AllowedHostsValidator:
		return self._allowed_hosts_validator
	
	async def get_authorization_token(
			self,
			uri: str,
			additional_authentication_context: Dict[str, Any] = {}) -> str:
		parsed_url = urlparse(uri)

		if not all([parsed_url.scheme, parsed_url.netloc]):
			return ""
		
		if not parsed_url.scheme == "https":
			return ""
		device_code_response = await self.get_device_code()
		print(f"Please go to {device_code_response.verification_uri} and enter the code {device_code_response.user_code} to authenticate.")
		if not device_code_response:
			return ""
		token = await self.poll_for_authorization_token(device_code_response)
		return token
		
	async def poll_for_authorization_token(self,
										device_code_response: GitHubDeviceCodeResponse) -> str:
		token = await self.get_token(device_code_response)
		while token == "":
			await asyncio.sleep(device_code_response.interval + 1)
			token = await self.get_token(device_code_response)
			if token != "":
				print("Authorized!")
				break
		
		return token
	async def get_token(self,
					device_code_response: GitHubDeviceCodeResponse) -> str:
		target_url = "https://github.com/login/oauth/access_token"
		request = httpx.Request("POST", target_url, headers= {
			"Content-Type": "application/x-www-form-urlencoded",
			"Accept": "application/json"
		}, data = {
			"client_id": self._client_id,
			"device_code": device_code_response.device_code,
			"grant_type": "urn:ietf:params:oauth:grant-type:device_code"
		})
		response = await httpx.AsyncClient().send(request)
		if response.status_code == 200:
			json_response = response.json()
			if json_response.get('access_token'):
				return json_response["access_token"]
		return ""

	async def get_device_code(self) -> GitHubDeviceCodeResponse:
		target_url = "https://github.com/login/device/code"
		request = httpx.Request("POST", target_url, headers= {
			"Content-Type": "application/x-www-form-urlencoded",
			"Accept": "application/json"
		}, data = {
			"client_id": self._client_id,
			"scope": self._scope
		})
		response = await httpx.AsyncClient().send(request)
		if response.status_code == 200:
			json_response = response.json()
			return GitHubDeviceCodeResponse(
				json_response["device_code"],
				json_response["user_code"],
				json_response["verification_uri"],
				int(json_response["expires_in"]),
				int(json_response["interval"]))
		return None

File: php/src/GitHubAuthentication/GitHubAccessTokenProvider.php
<?php

namespace Baywet\Githubtododemo\Githubauthentication;

use Http\Promise\FulfilledPromise;
use Http\Promise\Promise;
use Http\Promise\RejectedPromise;
use Microsoft\Kiota\Abstractions\Authentication\AccessTokenProvider;
use Microsoft\Kiota\Abstractions\Authentication\AllowedHostsValidator;
use GuzzleHttp\Client;

class GitHubAccessTokenProvider implements AccessTokenProvider
{
	private AllowedHostsValidator $allowedHostsValidator;
	private Client $client;
	private string $clientId;
	private string $scope;
	public function __construct(string $clientId, string $scope, array $allowedHosts = ["api.github.com"]) {
		$this->allowedHostsValidator = new AllowedHostsValidator();
        $this->allowedHostsValidator->setAllowedHosts($allowedHosts);
		$this->client = new Client();
		$this->clientId = $clientId;
		$this->scope = $scope;
	}

	public function getAllowedHostsValidator(): AllowedHostsValidator
    {
        return $this->allowedHostsValidator;
    }
	public function getAuthorizationTokenAsync(string $url, array $additionalAuthenticationContext = []): Promise
    {
        $parsedUrl = parse_url($url);
        $scheme = $parsedUrl["scheme"] ?? null;
        $host = $parsedUrl["host"] ?? null;
        try {
            if ($scheme !== 'https' || !$this->getAllowedHostsValidator()->isUrlHostValid($url)) {
                return new FulfilledPromise(null);
            }
			$deviceCodeResponse = $this->getDeviceCode();
			if(!$deviceCodeResponse) {
				return new RejectedPromise(new \Exception("Error getting device code"));
			}
			echo "Go to " . $deviceCodeResponse->verification_uri . " and enter " . $deviceCodeResponse->user_code . " to authenticate\n";
			$tokenResponse = $this->pollForToken($deviceCodeResponse);
			if(!$tokenResponse) {
				return new RejectedPromise(new \Exception("Error getting token"));
			}
			return new FulfilledPromise($tokenResponse->access_token);
		}
		catch (\Exception $ex) {
			return new RejectedPromise($ex);
		}
	}
	private function pollForToken(GitHubDeviceCodeResponse $deviceCodeResponse)
	{
		$result = $this->getToken($deviceCodeResponse);
		if ($result !== null) {
			return $result;
		} else {
			sleep($deviceCodeResponse->interval_in_seconds + 1);
			return $this->pollForToken($deviceCodeResponse);
		}

	}
	private function getToken(GitHubDeviceCodeResponse $deviceCodeResponse)
	{
		$response = $this->client->post("https://github.com/login/oauth/access_token", [
			"form_params" => [
				"client_id" => $this->clientId,
				"device_code" => $deviceCodeResponse->device_code,
				"grant_type" => "urn:ietf:params:oauth:grant-type:device_code"
			],
			"headers" => [
				"Accept" => "application/json"
			]
		]);
		if ($response->getStatusCode() !== 200) {
			throw new \Exception("Error getting token: " . $response->getStatusCode() . " " . $response->getBody());
		}
		$body = $response->getBody()->getContents();
		$result = GitHubAccessTokenCodeResponse::fromJson($body);
		if ("authorization_pending" === $result->error) {
			return null;
		} else if ($result->error !== "" && $result->error !== null) {
			throw new \Exception("Error getting token: " . $result->error . " " . $result->error_description);
		}
		return $result;
	}
	private function getDeviceCode()
	{
		$response = $this->client->post("https://github.com/login/device/code", [
			"form_params" => [
				"client_id" => $this->clientId,
				"scope" => $this->scope
			],
			"headers" => [
				"Accept" => "application/json"
			]
		]);
		if ($response->getStatusCode() !== 200) {
			throw new \Exception("Error getting device code: " . $response->getStatusCode() . " " . $response->getBody());
		}
		$body = $response->getBody()->getContents();
		return GitHubDeviceCodeResponse::fromJson($body);

	}
}